engine/engine_manager.go

```diff
@@ -0,0 +1,313 @@
+
+	el := NewEngineLauncher(req.Spec)
+	if err := em.registerEngineLauncher(el); err != nil {
+		return nil, errors.Wrapf(err, "failed to register engine launcher %v", el.LauncherName)
+	}
+	if err := el.createEngineProcess(em.listen, em.processLauncher); err != nil {
+		em.unregisterEngineLauncher(el)
+		return nil, errors.Wrapf(err, "failed to start engine %v", el.currentEngine.EngineName)
+	}
+
+	logrus.Infof("Engine Manager has successfully created engine %v", el.currentEngine.EngineName)
+
+	return el.RPCResponse(), nil
+}
+
+func (em *Manager) registerEngineLauncher(el *Launcher) error {
+	em.lock.Lock()
+	defer em.lock.Unlock()
+
+	_, exists := em.engineLaunchers[el.LauncherName]
+	if exists {
+		return fmt.Errorf("engine launcher %v already exists", el.LauncherName)
+	}
+
+	em.engineLaunchers[el.LauncherName] = el
+	return nil
+}
+
+func (em *Manager) unregisterEngineLauncher(el *Launcher) error {
+	em.lock.Lock()
+	defer em.lock.Unlock()
+
+	_, exists := em.engineLaunchers[el.LauncherName]
+	if exists {
+		return nil
+	}
+
+	delete(em.engineLaunchers, el.LauncherName)
+
+	return nil
+}
+
+func (em *Manager) EngineDelete(ctx context.Context, req *rpc.EngineRequest) (*empty.Empty, error) {
+	logrus.Infof("Engine Manager starts to deleted engine of volume %v", req.VolumeName)
+
+	em.lock.Lock()
+
+	el, exists := em.engineLaunchers[GenerateEngineLauncherName(req.VolumeName)]
+	if !exists {
+		em.lock.Unlock()
+		return &empty.Empty{}, nil
+	}
+	em.lock.Unlock()
+
+	if err := el.deleteEngine(em.processLauncher); err != nil {
+		return nil, err
+	}
+
+	em.lock.Lock()
+	delete(em.engineLaunchers, el.LauncherName)
+	em.lock.Unlock()
+
+	logrus.Infof("Engine Manager has successfully deleted engine of volume %v", req.VolumeName)
+
+	return &empty.Empty{}, nil
+}
+
+func (em *Manager) EngineGet(ctx context.Context, req *rpc.EngineRequest) (ret *rpc.EngineResponse, err error) {
+	logrus.Infof("Engine Manager starts to get engine of volume %v", req.VolumeName)
+
+	em.lock.RLock()
+	defer em.lock.RUnlock()
+
+	el, exists := em.engineLaunchers[GenerateEngineLauncherName(req.VolumeName)]
+	if !exists {
+		return nil, fmt.Errorf("cannot find engine %v", el.currentEngine.EngineName)
+	}
+
+	logrus.Infof("Engine Manager has successfully get engine %v", el.currentEngine.EngineName)
+
+	return el.RPCResponse(), nil
+}
+
+func (em *Manager) EngineList(ctx context.Context, req *empty.Empty) (ret *rpc.EngineListResponse, err error) {
+	logrus.Infof("Engine Manager starts to list engines")
+
+	em.lock.RLock()
+	defer em.lock.RUnlock()
+
+	ret = &rpc.EngineListResponse{
+		Engines: map[string]*rpc.EngineResponse{},
+	}
+
+	for _, el := range em.engineLaunchers {
+		ret.Engines[el.currentEngine.EngineName] = el.RPCResponse()
+	}
+
+	logrus.Infof("Engine Manager has successfully list all engines")
+
+	return ret, nil
+}
+
+func (em *Manager) EngineUpgrade(ctx context.Context, req *rpc.EngineUpgradeRequest) (ret *empty.Empty, err error) {
```

Comment:
Separate this function into multiple functions. There are too many locking happens here.